x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankTests.java

```diff
@@ -37,6 +37,20 @@ public class TextSimilarityRankTests extends ESSingleNodeTestCase {
      */
     public static class InvalidInferenceResultCountProvidingTextSimilarityRankBuilder extends TextSimilarityRankBuilder {
 
+        private boolean hasInvalidDocumentIndices = false;
```

Comment:
Kind of a nit, but I think this test would be more configurable and understandable if we could just pass in an integer value for for `top_n`. It more closely reflects how this error would occur in production.

Reply:
It does make things cleaner, but then we should also remove the "invalid indices in the top_n < rank_window_size" test case because the process doesn't care about the invalid indices, it raises the error beforehand (see [comment](https://github.com/elastic/elasticsearch/pull/111212#discussion_r1689985230)). WDYT?

Reply:
I don't think we should remove that test, but we should rename it. Ultimately, we need a test for when `top_n` < `rank_window_size`, no matter what it is called.

Reply:
Agreed, we should have a test that throws a `BAD_REQUEST` when we see that `rank_window_size` is larger than the referenced inference top_n configuration.

This check should be done eagerly, before the inference request is made.

Reply:
@Mikep86  What I mean is currently there are two overlapping tests: `testRerankInferenceResultCountMismatch` and `testRerankInvalidDocumentIndices`. Both set the `top_n` to be less than `rank_window_size`, the latter also corrupts some document indices in the reranker respose to point outside the returned range. This is what was throwing the IOOE the first place.

This state however will never be reached with this fix in place, because if `top_n < rank_window_size`, we fail fast without ever reaching the result processing and finding the invalid indices. So I don't see a point for keeping both test cases.

There _is_ a possible use case where `top_n == rank_window_size` and yet there is an `index` property pointing outside the N elements. But this means there's something broken in the reranker and it doesn't honor the API contract, which is not under our control. I'd argue whether we want to add new code to validate the result and throw an exception if the indices are off, or just let the IOOE cover it. (If we did that, where do we draw the line for validation? Checking that they are all integers? Ensuring there are no duplicates? etc.)

Reply:
Got it, then i don't see a reason to keep the second test. Let's modify `testRerankInferenceResultCountMismatch` to check that `BAD_REQUEST` is returned, per @benwtrent's guidance.